### PR TITLE
UCP: request.send refactoring

### PR DESCRIPTION
**Pull request by evgeny-leksikov**
- ~~unite send.state (uct_comp + dt)~~ will be in other PR
 - remove ugly logic around saved_state in bcopy/zcopy
 - add ucp_request_send_state_set and ucp_request_send_state_advance
   functions for generic send.state management
 - rename ucp_tag_req_start to ucp_request_send_start and reuse it
   for stream request, it adds IOVs support to ucp_stream_send_nb
 - rename ucp_request_start_send to ucp_request_send because of name
   conflict with above
**Update:**
 - merged start generic with start contig/iov

testing of stream API + IOVs is in next PR #1875

### CODE REVIEW

File: src/ucp/core/ucp_request.c [hunk 1]
```diff
@@ -296,3 +299,112 @@ ucs_status_t ucp_request_test(void *request, ucp_tag_recv_info_t *info)
...
+        ucp_request_send_state_set(req, 0, 0, UCP_REQUEST_SEND_STATE_RNDV);
+        status = ucp_tag_send_start_rndv(req);
+        if (status != UCS_OK) {
+            return status;
+        }
+        UCS_PROFILE_REQUEST_EVENT(req, "start_rndv", req->send.length);
+    } else if (length < zcopy_thresh) {
+        /* bcopy */
+        ucp_request_send_state_set(req, 0, 0, UCP_REQUEST_SEND_STATE_BCOPY);
+        if (length <= (config->tag.eager.max_bcopy - only_hdr_size)) {
```
Comment (evgeny-leksikov) [inline src/ucp/core/ucp_request.c]: @yosefe maybe need to add thresholds to `proto` structure and initialize it on startup according to config?

File: src/ucp/core/ucp_request.c [hunk 2]
```diff
@@ -296,3 +299,112 @@ ucs_status_t ucp_request_test(void *request, ucp_tag_recv_info_t *info)
...
+
+    if (((ssize_t)length <= max_short) && !is_iov) {
+        /* short */
+        req->send.uct.func = proto->contig_short;
+        UCS_PROFILE_REQUEST_EVENT(req, "start_contig_short", req->send.length);
+    } else if ((length >= rndv_rma_thresh) || (length >= rndv_am_thresh) ||
+               force_sw_rndv) {
+        /* RMA/AM rendezvous */
+        ucp_request_send_state_set(req, 0, 0, UCP_REQUEST_SEND_STATE_RNDV);
+        status = ucp_tag_send_start_rndv(req);
```
Comment (evgeny-leksikov) [inline src/ucp/core/ucp_request.c]: Stream API is not support RNDV yet, i will add the comment, need to add `start_rndv` to `proto`.
looks like it's time for https://github.com/openucx/ucx/pull/1436

File: src/ucp/core/ucp_request.c (hunk 1 above)
Comment (evgeny-leksikov) [reply src/ucp/core/ucp_request.c]: Instead of header sizes. maybe i missed something but we are using header sizes directly where they are really needed.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4716/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4717/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4719/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
bot:bgate:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4731/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4735/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4736/ for details (Mellanox internal link).

File: src/ucp/core/ucp_ep.c [hunk 3]
```diff
@@ -489,12 +490,13 @@ void ucp_ep_err_pending_purge(uct_pending_req_t *self, void *arg)
     ucp_request_t *req      = ucs_container_of(self, ucp_request_t, send.uct);
     ucs_status_t  status    = UCS_PTR_STATUS(arg);
 
-    if (req->send.uct_comp.func) {
-        req->send.state.offset = req->send.length; /* fast-forward to data end */
+    if (req->send.state.uct_comp.func) {
+        /* TODO: use set state */
```
Comment (evgeny-leksikov) [inline src/ucp/core/ucp_ep.c]: add `ucp_request_send_state_ff` (fast-forward)

File: src/ucp/tag/tag_send.c [hunk 4]
```diff
@@ -11,183 +11,82 @@
...
+    case UCP_DATATYPE_CONTIG: 
+        return ucs_min(rndv_rma_thresh, rndv_am_thresh);
+    case UCP_DATATYPE_GENERIC:
+        return rndv_am_thresh;
+    default:
+        ucs_error("Invalid data type %lx", req->send.datatype);
     }
-    return UCS_OK;
+ 
+    return SIZE_MAX; /* TODO: check */
```
Comment (evgeny-leksikov) [inline src/ucp/tag/tag_send.c]: remove TODO comment

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4745/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4746/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4747/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4748/ for details (Mellanox internal link).

File: src/ucp/core/ucp_request.c [hunk 5]
```diff
@@ -296,3 +298,54 @@ ucs_status_t ucp_request_test(void *request, ucp_tag_recv_info_t *info)
     }
     return UCS_INPROGRESS;
 }
+
+ucs_status_t
+ucp_request_send_start(ucp_request_t *req, ssize_t max_short,
+                       size_t zcopy_thresh, size_t seg_size,
+                       size_t rndv_thresh, const ucp_proto_t *proto)
```
Comment (yosefe) [inline src/ucp/core/ucp_request.c]: s/rndv_thresh/max_zcopy

File: src/ucp/core/ucp_request.h [hunk 6]
```diff
@@ -142,14 +152,14 @@ struct ucp_request {
 
             };
 
-            /* TODO: to unite state and uct_comp in single req_state and
-             *       implement common functions to work with it,
-             *       See conversation to PR #1737 */
+            struct {
```
Comment (yosefe) [inline src/ucp/core/ucp_request.h]: that change can be in another PR (to be under 500 lines)

File: src/ucp/proto/proto_am.inl [hunk 7]
```diff
@@ -194,85 +175,125 @@ ucs_status_t ucp_do_am_zcopy_multi(uct_pending_req_t *self, uint8_t am_id_first,
...
+                                   UCP_REQUEST_SEND_PROTO_ZCOPY_AM,
+                                   status);
+
+    return (status < 0) ? status : UCS_INPROGRESS;
+}
+
+static UCS_F_ALWAYS_INLINE size_t
+ucp_proto_get_zcopy_threshold(const ucp_request_t *req,
+                              const ucp_ep_msg_config_t *msg_config,
+                              size_t count, size_t rndv_thresh)
```
Comment (yosefe) [inline src/ucp/proto/proto_am.inl]: s/rndv_thresh/max_zcopy

File: src/ucp/core/ucp_request.h [hunk 8]
```diff
@@ -45,7 +45,17 @@ enum {
 };
 
 
- /**
+/**
+ * Protocols enumerator to work with send request state
+ */
+enum {
+    UCP_REQUEST_SEND_PROTO_BCOPY_AM = 0,
```
Comment (yosefe) [inline src/ucp/core/ucp_request.h]: :+1:

File: src/ucp/tag/tag_send.c [hunk 9]
```diff
@@ -11,183 +11,81 @@
...
-        UCS_PROFILE_REQUEST_EVENT(req, "start_gen_bcopy_single", req->send.length);
-    } else if (length >= rndv_am_thresh) {
-        /* rendezvous */
-        ucp_tag_send_start_rndv(req);
-        UCS_PROFILE_REQUEST_EVENT(req, "start_rndv", req->send.length);
-    } else {
-        /* bcopy multi */
-        req->send.uct.func = proto->bcopy_multi;
-        UCS_PROFILE_REQUEST_EVENT(req, "start_gen_bcopy_multi", req->send.length);
+    if (ucs_unlikely(UCP_DT_IS_IOV(datatype))) {
```
Comment (yosefe) [inline src/ucp/tag/tag_send.c]: return ucs_likely(!IS_IOV(datatype)) || (count <= max_iov)

File: src/ucp/tag/tag_send.c [hunk 10]
```diff
@@ -11,183 +11,81 @@
...
                  ucp_send_callback_t cb, const ucp_proto_t *proto)
 {
-    ucs_status_t status;
-
-    switch (req->send.datatype & UCP_DATATYPE_CLASS_MASK) {
-    case UCP_DATATYPE_CONTIG:
-    case UCP_DATATYPE_IOV:
-        status = ucp_tag_req_start(req, count, max_short, zcopy_thresh,
-                                   rndv_rma_thresh, rndv_am_thresh, proto);
+    int flag_iov_single = ucp_tag_is_iov_single(req->send.datatype, count,
```
Comment (yosefe) [inline src/ucp/tag/tag_send.c]: looks like the name 'iov_single' was wrong in the first place, since it compares to msg_config->max_iov and not 1.
also, why not put the code of this function inside ucp_tag_get_rndv_threshold()?

Comment (evgeny-leksikov) [reply src/ucp/tag/tag_send.c]: 👍 this variable was used in several places in original code. i have not noticed as simplified it during refactoring :)

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4755/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4756/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4760/ for details (Mellanox internal link).

File: src/ucp/core/ucp_request.inl [hunk 11]
```diff
@@ -165,6 +165,124 @@ void ucp_request_recv_generic_dt_finish(ucp_request_t *req)
...
+    }
+
+    /* offset is not used for RMA */
+    ucs_assert((proto == UCP_REQUEST_SEND_PROTO_RMA) ||
+               (req->send.state.offset <= req->send.length));
+}
+
+static UCS_F_ALWAYS_INLINE void
+ucp_request_send_state_advance(ucp_request_t *req,
+                               const ucp_dt_state_t *dt_state, unsigned proto,
```
Comment (yosefe) [inline src/ucp/core/ucp_request.inl]: code style - pls arrange parameters so single line will be as long as possible (but up to 80)
also rename dt_state to "new_dt_state" and add comment to explain this function usage

File: src/ucp/proto/proto_am.inl [hunk 12]
```diff
@@ -194,85 +175,125 @@ ucs_status_t ucp_do_am_zcopy_multi(uct_pending_req_t *self, uint8_t am_id_first,
...
+    }
+
+    ucp_request_send_state_advance(req, &state,
+                                   UCP_REQUEST_SEND_PROTO_ZCOPY_AM,
+                                   status);
+
+    return (status < 0) ? status : UCS_INPROGRESS;
+}
+
+static UCS_F_ALWAYS_INLINE size_t
```
Comment (yosefe) [inline src/ucp/proto/proto_am.inl]: i think this function should not be inline, zcopy is not fast path
can we somehow arrange that this function will be called only if short is not used? maybe add "pure" attribute?

File: src/ucp/core/ucp_request.inl [hunk 13]
```diff
@@ -165,6 +165,124 @@ void ucp_request_recv_generic_dt_finish(ucp_request_t *req)
...
+        req->send.state.dt.generic.state = state_gen;
+        return;
+    default:
+        ucs_error("Invalid data type");
+    }
+}
+
+static UCS_F_ALWAYS_INLINE void
+ucp_request_send_state_reset(ucp_request_t *req,
+                             uct_completion_callback_t comp_cb, unsigned proto)
```
[the remark this answers is not in the document]
Comment (evgeny-leksikov) [reply src/ucp/core/ucp_request.inl]: moving `uct_completion_callback_t comp_cb` to line before make its length 84.

File: src/ucp/core/ucp_request.inl (hunk 11 above)
Comment (evgeny-leksikov) [reply src/ucp/core/ucp_request.inl]: the same: > 80. i will rename `dt_state` arg and add comment

File: src/ucp/proto/proto_am.inl (hunk 12 above)
Comment (evgeny-leksikov) [reply src/ucp/proto/proto_am.inl]: I measured performance of this PR, no effects on osu_latency and osu_mbw_mr. if we will add check for `short_max` outside this function it will look ugly. I think doing the calculations of all thresholds on initialization and save them in proto structures instead of header lengths would be better. the only one exception is iovs where we have dependency from input data type/len/count. It would be just usual function under `unlikely(is_iov)` branch.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4781/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4783/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4784/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4785/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
:+1: 
can you please squash?

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4818/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4819/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
@yosefe it looks ready to go
